be/src/olap/olap_server.cpp

```diff
@@ -1073,6 +1073,9 @@ Status StorageEngine::_submit_compaction_task(TabletSharedPtr tablet,
                              "task, tablet_id="
                           << tablet->tablet_id() << ", tablet_state=" << tablet->tablet_state();
                 _pop_tablet_from_submitted_compaction(tablet, compaction_type);
+                if (!force) {
+                  _permit_limiter.release(permits);
+                }
                 return;
```

Comment:
we should set
```
tablet->compaction_stage = CompactionStage::NOT_SCHEDULED;
```
as well

Reply:
And use defer to make it better:
Defer defer {
    [&]() {
        _pop_tablet_from_submitted_compaction(tablet, compaction_type);
        tablet->compaction_stage = CompactionStage::NOT_SCHEDULED;
    }
};

And another one in if block:
Defer defer { 
      if (!force) {
           _permit_limiter.release(permits);
      } 
}